services/src/main/java/org/keycloak/services/clienttype/client/TypeAwareClientModelDelegate.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.keycloak.services.clienttype.client;
+
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+import org.keycloak.common.util.ObjectUtil;
+import org.keycloak.models.ClientModel;
+import org.keycloak.models.delegate.ClientModelLazyDelegate;
+import org.keycloak.client.clienttype.ClientType;
+import org.keycloak.client.clienttype.ClientTypeException;
+
+/**
+ * Delegates to client-type and underlying delegate
+ *
+ * @author <a href="mailto:mposolda@redhat.com">Marek Posolda</a>
+ */
+public class TypeAwareClientModelDelegate extends ClientModelLazyDelegate {
+
+    private final ClientType clientType;
+
+    public TypeAwareClientModelDelegate(ClientType clientType, Supplier<ClientModel> clientModelSupplier) {
+        super(clientModelSupplier);
+
+        if (clientType == null) {
+            throw new IllegalArgumentException("Null client type not supported for client " + getClientId());
+        }
+        this.clientType = clientType;
+    }
+
+    @Override
+    public boolean isStandardFlowEnabled() {
+        return getBooleanProperty("standardFlowEnabled", super::isStandardFlowEnabled);
+    }
+
+    @Override
+    public void setStandardFlowEnabled(boolean standardFlowEnabled) {
+        setBooleanProperty("standardFlowEnabled", standardFlowEnabled, super::setStandardFlowEnabled);
+    }
+
+
+    protected boolean getBooleanProperty(String propertyName, Supplier<Boolean> clientGetter) {
+        // Check if clientType supports the feature. If not, simply return false
+        if (!clientType.isApplicable(propertyName)) {
```

Comment:
Not sure if this is appropriate here. Like if the property is not applicable for the client type, should we throw an error instead of returning false? Same with in setBooleanProperty.

Reply:
Yeah sorry I was more opening up this for discuss and hadn't really looked into this too far. Like why return false here? Why not null and use return type Boolean? Or even return the value from the delegate?

I think this class needs to be discussed in general. See comment here, if you have some insights - https://github.com/keycloak/keycloak/discussions/9066#discussioncomment-9000803

Reply:
@patrickjennings I think returning `null` instead of `false` for non-applicable properties would be really great. IMO it will be cleaner than returning `false`. However I think we are restricted by the signature of methods on `ClientModel` . For example method `ClientModel.isStandardFlowEnabled()` (as well as all other "boolean" getter/setter methods on `ClientModel`) are using just simple `boolean` . So I think returning null unfortunately won't work due to this...

Returning the value from the delegate: I am not seeing much point in this if the property is "non-applicable" for this client type. Like for instance if client type would be `service account`, then properties like `alwaysDisplayInConsole` are not applicable for this client type as normal users cannot login to those clients and hence cannot display them in account console. So since property is just "not applicable" for those clients, we can automatically return `false` for property `alwaysDisplayInConsole` for all clients of client type `service account` due to this without a need to try delegate.